pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,262 @@
+			appNum, appNumOnUNetKey(baseID, appID))
+		baseMap.Set(appNum)
+		// Clear InUse
+		uuidpairtonum.NumFree(log, pub, baseID, appID)
+	}
+	// In case zedrouter process restarted we fill in InUse from
+	// AppNetworkStatus, underlay network entries
+	items = pubAppNetworkStatus.GetAll()
+	for _, item := range items {
+		status := item.(types.AppNetworkStatus)
+		appID := status.UUIDandVersion.UUID
+
+		// If we have a config for the UUID we should mark it as
+		// allocated; otherwise mark it as reserved.
+		// XXX however, on startup we are not likely to have any
+		// config yet.
+		for i := range status.UnderlayNetworkList {
+			ulStatus := &status.UnderlayNetworkList[i]
+			baseID := ulStatus.Network
+			baseMap := appNumOnUNetBaseGet(baseID)
+			appNum, err := uuidpairtonum.NumGet(log, pub,
+				baseID, appID, numType)
+			if err != nil {
+				continue
+			}
+			if !baseMap.IsSet(appNum) {
+				log.Fatalf("Bitmap is not set for %s num %d",
+					appNumOnUNetKey(baseID, appID), appNum)
+			}
+			log.Functionf("Marking InUse for appNum %d",
+				appNum)
+			// Set InUse
+			uuidpairtonum.NumAllocate(log, pub, baseID,
+				appID, appNum, false, numType)
+		}
+	}
+}
+
+// If an entry is not inUse and and its CreateTime were
+// before the agent started, then we free it up.
+func appNumMapOnUNetGC(ctx *zedrouterContext) {
+
+	pub := ctx.pubUUIDPairToNum
+	numType := appNumOnUNetType
+
+	log.Functionf("appNumOnUNetMapGC")
+	freedCount := 0
+	items := pub.GetAll()
+	for _, item := range items {
+		appNumMap := item.(types.UUIDPairToNum)
+		if appNumMap.NumType != numType {
+			continue
+		}
+		if appNumMap.InUse {
+			continue
+		}
+		if appNumMap.CreateTime.After(ctx.agentStartTime) {
+			continue
+		}
+		log.Functionf("appNumMapOnUNetGC: freeing %+v", appNumMap)
+		appNumOnUNetFree(ctx, appNumMap.BaseID, appNumMap.AppID)
+		freedCount++
+	}
+	log.Functionf("appNumMapOnUNetGC freed %d", freedCount)
+}
+
+func appNumOnUNetAllocate(ctx *zedrouterContext, baseID uuid.UUID,
+	appID uuid.UUID, isStatic bool, isZedmanager bool) (int, error) {
+
+	pub := ctx.pubUUIDPairToNum
+	numType := appNumOnUNetType
+	baseMap := appNumOnUNetBaseCreate(baseID)
+
+	// Do we already have a number?
+	appNum, err := uuidpairtonum.NumGet(log, pub, baseID, appID,
+		numType)
+	if err == nil {
+		log.Functionf("Found allocated appNum %d for %s", appNum,
+			appNumOnUNetKey(baseID, appID))
+		if !baseMap.IsSet(appNum) {
+			log.Fatalf("Bitmap value(%d) is not set", appNum)
+		}
+		// Set InUse and update time
+		uuidpairtonum.NumAllocate(log, pub, baseID, appID, appNum,
+			false, numType)
+		return appNum, nil
+	}
+
+	// Find a free number in bitmap; look for zero if isZedmanager
+	if isZedmanager && !baseMap.IsSet(0) {
+		appNum = 0
+		log.Functionf("Allocating appNum %d for %s isZedmanager",
+			appNum, appNumOnUNetKey(baseID, appID))
+	} else {
+		// XXX could look for non-0xFF bytes first for efficiency
+		appNum = -1
+		if isStatic {
```

Comment:
It would make sense to add a comment that for static we pick the topmost numbers so avoid consuming dynamic IP address from a smallish DHCP range.